opm/simulators/linalg/ISTLSolverEbos.hpp

```diff
@@ -785,103 +375,74 @@ DenseMatrix transposeDenseMatrix(const DenseMatrix& M)
-
-            std::vector<std::set<size_type>> pattern;
-            pattern.resize(numCells);
 
-            auto elemIt = gridView.template begin<0>();
-            const auto& elemEndIt = gridView.template end<0>();
-
-            //Loop over cells
-            for (; elemIt != elemEndIt; ++elemIt)
-            {
-                const auto& elem = *elemIt;
-                size_type idx = elemMapper.index(elem);
-                pattern[idx].insert(idx);
-
-                // Add well non-zero connections
-                for (auto wc = wellConnectionsGraph_[idx].begin(); wc!=wellConnectionsGraph_[idx].end(); ++wc)
-                    pattern[idx].insert(*wc);
-
-                // Add just a single element to ghost rows
-                if (elem.partitionType() != Dune::InteriorEntity)
-                {
-                    noGhostMat_->setrowsize(idx, pattern[idx].size());
-                }
-                else {
-                    auto isend = gridView.iend(elem);
-                    for (auto is = gridView.ibegin(elem); is!=isend; ++is)
-                    {
-                        //check if face has neighbor
-                        if (is->neighbor())
-                        {
-                            size_type nid = elemMapper.index(is->outside());
-                            pattern[idx].insert(nid);
-                        }
-                    }
-                    noGhostMat_->setrowsize(idx, pattern[idx].size());
+        /// Return true if we should (re)create the whole solver,
+        /// instead of just calling update() on the preconditioner.
+        bool shouldCreateSolver() const
+        {
+            // Decide if we should recreate the solver or just do
+            // a minimal preconditioner update.
+            if (!flexibleSolver_) {
+                return true;
+            }
+            if (this->parameters_.cpr_reuse_setup_ == 0) {
+                // Always recreate solver.
+                return true;
+            }
+            if (this->parameters_.cpr_reuse_setup_ == 1) {
+                // Recreate solver on the first iteration of every timestep.
+                const int newton_iteration = this->simulator_.model().newtonMethod().numIterations();
+                if (newton_iteration == 0) {
+                    return true;
                 }
             }
-            noGhostMat_->endrowsizes();
-            for (size_type dofId = 0; dofId < numCells; ++dofId)
-            {
-                auto nabIdx = pattern[dofId].begin();
-                auto endNab = pattern[dofId].end();
-                for (; nabIdx != endNab; ++nabIdx)
-                {
-                    noGhostMat_->addindex(dofId, *nabIdx);
+            if (this->parameters_.cpr_reuse_setup_ == 2) {
+                // Recreate solver if the last solve used more than 10 iterations.
+                if (this->iterations() > 10) {
+                    return true;
                 }
```

Comment:
Writing it this way will trigger the `assert` at the end of the function if `this->iteration() <= 10`. You really need an unconditional return here&mdash;or remove the `assert`&mdash;if you don't want that to happen.

Reply:
Simplified and fixed (I think).

Reply:
> Simplified and fixed (I think).

Yep, looks good now.  I vote to merge this once the build check is green but will let others speak to this too.